What are container components in React. (#2229)
section-engineering-education/engineering-education: articles/container-components-in-react/index.md

### What are Container Components in React.

As time goes on, while making use of React, one would need to think of how data is being passed among different components in a React application.

> In this article we’d create a search component that makes use of the container component React pattern.

Author: Oh!

I'd get an elaborate definition.

When you’re done reading this article, you’d have an understanding of how a container component works and how to build any React component using this pattern.

### What you need to know

> Before you read this article, you need to have an idea of what [React](https://reactjs.org) is, how to make use of [React Hooks](https://reactjs.org/docs/hooks-intro.html) to store the state in a component, and understand how components and [Props](https://reactjs.org/docs/components-and-props.html) work together.

Author: Noted!

```js
import React from “react”

export const exampleComponent = ({ children, open }) => {
	return (
		<div onClick={() => open()}>
			{children}
		</div>
	);
};

```

The component above illustrates how a container component performs two functions, presenting the UI and handling the logic of the component.

If you look at the component above, you’d see that it makes use of props to share data among components in the “arbitrary” React app.

Note: There’s a convention that is quite common in the developer community.

Developers tend to place their container components in a separate folder and the UI components into another folder, so they can distinguish between these components. You can choose not to follow this pattern of creating different folders if you already have one that works fine for you, so long you’re able to solve the same problem.

Building the search component
A search component is among the most ubiquitous components in a typical React application.

Let’s start by thinking about the function(s) that the search component would perform. A search component should be a able to perform basic tasks like searching and filtering.

> To limit the amout of time spent on getting the perfect API that the search component has to interact with.

Author: Awesome, I'd do that.

Thanks

> I’d go ahead to create an array of objects containing simple information that will be in this component

Author: Thank you for this suggestion. I'd go on to apply it now.

If you look at the component above, you’d see that it makes use of props to share data among components in the “arbitrary” React app.

**Note:** There’s a convention that is quite common in the developer community.

Developers tend to place their container components in a separate folder and the UI components into another folder, so they can distinguish between these components. You can choose not to follow this pattern of creating different folders if you already have one that works fine for you, so long you’re able to solve the same problem.

### Building the search component

A search component is among the most ubiquitous components in a typical React application.

Let’s start by thinking about the function(s) that the search component would perform. A search component should be a able to perform basic tasks like searching and filtering.

To keep things simple, let's use an array of objects as the data to search from, instead of using an API.

Let's create an array of objects containing simple information about books that will be in this component:

```js
// bookshelf.js

> export const bookShelf = [
>   {
>     bookName: "Eloquent JavaScript",
>     author: "Marijn Haverbeke",
>   },
>   {
>     bookName: "CSS The Definitive Guide",
>     author: "Eric A. Meyer & Estelle Weyl",
>   },
>   {
>     bookName: "Automate the boring stuff with Python",
>     author: "Al Sweigart",
>   },
>   {
>     bookName: "Make: JavaScript Robotics",
>     author: "Donovan Buck, Bryan Hughes et al",
>   },
> ];

Reviewer: ```suggestion
export const books = [
  {
    name: "Eloquent JavaScript",
    author: "Marijn Haverbeke",
  },
  {
    name: "CSS The Definitive Guide",
    author: "Eric A. Meyer & Estelle Weyl",
  },
  {
    name: "Automate the boring stuff with Python",
    author: "Al Sweigart",
  },
  {
    name: "Make: JavaScript Robotics",
    author: "Donovan Buck, Bryan Hughes et al",
  },
];
```

```

> The data in `authors.js` will serve as the information that would be mapped onto the `authorsPage` so the search component can be able to filter the list of authors once the user starts typing into the input field.

Author: > Can you be more clear. Where did this `author.js` file come from? You haven't mentioned it yet and it's a bit confusing.

That is a typographical error, I've rectified it though,

Let's create the components needed in this app below and I'll explain how everything works.

The `Search` component presents the UI (i.e. It renders an input element where users can filter the books in the list.) to the user. It makes use of React props to pass data down to other components such as the author’s page container component.

```js
// search.js
import React from “react”

const Search = ({ val, onSearch, ...props }) => {
  return (
    <input
	  type="text"
	  value={val}
	  placeholder="Search for any book"
	  onChange={onSearch}
	  {...props}
	/>
  );
};

export default Search
```

The `Books` component performs almost the same function with the search component. Its only difference has to with the usage of React props to share data amongst the other components in the app.

Take a look at how the `bookShelf` component is being imported into the authors container and how it is being used in the `Books` component.

```js
// books.js
const Books = ({ books }) => {
  return (
  {books.map((bookShelf, index) => {
    return (
	  <ul key={index}>
	    <li>Book: {bookShelf.bookName}</li>
		<li>Author: {bookShelf.author}</li>
	  </ul>
    )
  })}
  )
}

> export default Books

Author: Uhm...

That wouldn't be necessary actually, since I'm just passing `bookShelf` as a callback variable/array that comes from the map function.

I've added the import statement of `books` from the `bookShelf.js` file in the appropriate component BTW.

```

The `authorsPage` component illustrates how a container component can perform two functions, rendering the UI and housing the logic of that container.

Notice how the usage of React Hooks is being used to store/hold the state of the component, which also is among the features of a container component, they are usually stateful components.

> The text that user types into the input field is being stored in an array so that we’d be able to perform a filtering functionality using [JavaScript’s filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/) method.

Reviewer: The `e.target.value` will provide a string from the input, not an array. You're setting the initial state of searchTerm as an empty string instead of an empty array.

Author: Wow!

How did I not notice this 😢

Reviewer: Hi @Caleb335. The line still mentions the text is being stored in an array. Just say it's stored in the state as a string.

This is possible because the state of the `Book` component is also monitored in this container component.

So each time the input field detects a character, it starts looping through and all the items in the bookShelf array to see which one matches the text and renders it to the DOM.

```js

> import React from “react”

Reviewer: ```suggestion
import React, { useState } from “react”
```

> import { bookShelf } from “./bookshelf.js”

Reviewer: ```suggestion
import { books } from “./bookshelf.js”
```

import Books from “./books”

const authorsContainer = () => {

>   const [searchTerm, setSearchTerm] = React.useState([]);

Reviewer: ```suggestion
  const [searchTerm, setSearchTerm] = useState("");
```

  const [books, setBooks] = React.useState(bookShelf);

  const handleInputChange = (e) => {
    setSearchTerm(e.target.value);
  };

  // perfoms a filtering on the questions
  // based on the users input and returns a new array
  // which gets rendered to the DOM.

>   const filteredBooks = books.filter((bookShelf) => {
>     return bookShelf.bookName.includes(searchTerm);
>   });

Reviewer: The variable names got really confusing at this point.

```suggestion
  const filteredBooks = bookShelf.filter((book) => {
    return book.bookName.includes(searchTerm);
  });
```

    author: "Donovan Buck, Bryan Hughes et al",
  },
];
```

The data in the `bookshelf.js` file will serve as the information. The search component should be able to filter the list of books once the user starts typing into the input field.

Let's create the components needed in this app below and I'll explain how everything works.

The `Search` component presents the UI (i.e. It renders an input element where users can filter the books in the list.) to the user.

```js
// search.js
import React from “react”

const Search = ({ val, onSearch, ...props }) => {
  return (
    <input
	  value={val}
	  onChange={onSearch}
	  placeholder="Search for any book"
	  {...props}
    />
  );
};

export default Search
```

The `Books` component performs almost the same function as the search component. Most people refer to this type of components as **"dumb", "stateless" or "presentational"** components because they're only concerned with how the user interface of the application is being rendred or displayed to the user. They do not always have state, instead they use props from parent component to get data from above in the application component tree.

```js

> // books.js
> const Books = ({ books }) => {
>   return (
>   {books.map((book, index) => {
>     return (
> 	  <ul key={index}>
> 	    <li>Book: {book.bookName}</li>
> 		<li>Author: {book.author}</li>
> 	  </ul>
>     )
>   })}
>   )
> }
> export default Books

Reviewer: ```suggestion
// books.js

import React from "react";

const Books = ({ books }) => {
  return books.map((book, index) => {
    return (
      <ul key={index}>
        <li>Book: {book.name}</li>
        <li>Author: {book.author}</li>
      </ul>
    );
  });
};
export default Books;

```

```

The `App` component illustrates how a container component can perform two functions, rendering the UI and housing the logic of that container.

The text that user types into the input field is stored in the `App` component's state as a string so that we’d be able to perform a filter the books using the [filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/) method.

This is possible because the data passed to the `Book` component is also controlled in this container component.

So each time the input field detects a character, it starts looping through and all the items in the `books` array to see which one matches the text and renders it to the DOM.

```js

> import React, { useState } from “react”
> import { books } from “./bookshelf.js”
> import Books from “./books”
> 
> const App = () => {
>   const [searchTerm, setSearchTerm] = useState("");
> 
>   const handleInputChange = (e) => {
>     setSearchTerm(e.target.value);
>   };
> 
>   // perfoms a filtering on the questions
>   // based on the users input and returns a new array
>   // which gets rendered to the DOM.
>   const filteredBooks = books.filter((book) => {
>     return book.bookName.includes(searchTerm);
>   });
> 
>   return (
>     <section>
> 	  <Search
> 	    name="search"
> 		onSearch={handleInputChange} val={searchTerm}
> 	  />
> 	  <Book books={filteredBooks}/>
> 	</section>
>   )
> }
> export default App;
> 

Reviewer: ```suggestion
import React, { useState } from "react";
import { books } from "./bookshelf.js";
import Books from "./Books";
import Search from './Search';

const App = () => {
  const [searchTerm, setSearchTerm] = useState("");
  const handleInputChange = (e) => {
    setSearchTerm(e.target.value);
  };
  // perfoms a filtering on the questions
  // based on the users input and returns a new array
  // which gets rendered to the DOM.

  const filteredBooks = books.filter((book) => {
    return book.name.includes(searchTerm);
  });
  return (
    <section>
      <Search name="search" onSearch={handleInputChange} val={searchTerm} />
      <Books books={filteredBooks} />
    </section>
  );
};
export default App;
```

```
src
 |-- App.js
 |-- bookshelf.js
 |-- components
           |---- Books.js
           |---- Search.js
```

Let’s start by thinking about the function(s) that the search component would perform. A search component should be a able to perform basic tasks like searching and filtering.

To keep things simple, let's use an array of objects as the data to search from, instead of using an API.

Let's create an array of objects containing simple information about books that will be in this component:

```js
// bookshelf.js
export const books = [
  {
    name: "Eloquent JavaScript",
    author: "Marijn Haverbeke",
  },
  {
    name: "CSS The Definitive Guide",
    author: "Eric A. Meyer & Estelle Weyl",
  },
  {
    name: "Automate the boring stuff with Python",
    author: "Al Sweigart",
  },
  {
    name: "Make: JavaScript Robotics",
    author: "Donovan Buck, Bryan Hughes et al",
  },
];
```

> The data in the `bookshelf.js` file will serve as the information that would be mapped onto the `App` component, so the search component can be able to filter the list of authors once the user starts typing into the input field.

Reviewer: ```suggestion
The data in the `bookshelf.js` file will serve as the information that would be mapped onto the `App` component, so the search component can be able to filter the list of books once the user starts typing into the input field.
```

Let's create the components needed in this app below and I'll explain how everything works.

The `Search` component presents the UI (i.e. It renders an input element where users can filter the books in the list.) to the user. It makes use of React props to pass data down to other components such as the author’s page container component.

```js
// search.js
import React from “react”

const Search = ({ val, onSearch, ...props }) => {
  return (
    <input
	  type="text"
	  value={val}
	  placeholder="Search for any book"
	  onChange={onSearch}
	  {...props}
	/>
  );
};

export default Search
```

The `Books` component performs almost the same function as the search component. Most people refer to this type of components as **"dumb", "stateless" or "presentational"** components because they're only concerned with how the user interface of the application is being rendred or displayed to the user. They do not always have state, instead they use props to share data in the application component tree.

Take a look at how the `bookShelf` component is being imported into the authors container and how it is being used in the `Books` component.

```js
// books.js

> const Books = ({ books }) => {
>   return (
>   {books.map((bookShelf, index) => {
>     return (
> 	  <ul key={index}>
> 	    <li>Book: {bookShelf.bookName}</li>
> 		<li>Author: {bookShelf.author}</li>
> 	  </ul>
>     )
>   })}
>   )
> }

Reviewer: ```suggestion
const Books = ({ books }) => {
  return (
  {books.map((book, index) => {
    return (
	  <ul key={index}>
	    <li>Book: {book.bookName}</li>
		<li>Author: {book.author}</li>
	  </ul>
    )
  })}
  )
}
```

Reviewer: Since you've already named the file bookshelf, rename these as `book` to prevent any sort of confusion.

Author: Ok...

That sounds good

This is possible because the state of the `Book` component is also monitored in this container component.

So each time the input field detects a character, it starts looping through and all the items in the bookShelf array to see which one matches the text and renders it to the DOM.

Notice how the usage of React Hooks is being used to store/hold the state of the component in the snippet below, which also is among the features of a container component, they are usually stateful components

```js
import React, { useState } from “react”
import { books } from “./bookshelf.js”
import Books from “./books”

const App = () => {
  const [searchTerm, setSearchTerm] = useState("");

  const handleInputChange = (e) => {
    setSearchTerm(e.target.value);
  };

  // perfoms a filtering on the questions
  // based on the users input and returns a new array
  // which gets rendered to the DOM.
  const filteredBooks = books.filter((book) => {
    return book.bookName.includes(searchTerm);
  });

  return (
    <section>
	  <Search
	    name="search"
		onSearch={handleInputChange} val={searchTerm}
	  />
	  <Book books={filteredBooks}/>
	</section>
  )
}

> export default authorsContainer

Reviewer: ```suggestion
export default App;
```

Author: > Hi @Caleb335. The line still mentions the text is being stored in an array. Just say it's stored in the state as a string.

Noted, I'd change that now.
